// @ckb-ccc/core/dist.commonjs/signer/nostr/signerNostrPrivateKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignerNostrPrivateKey = void 0;
const secp256k1_1 = require("@noble/curves/secp256k1");
const index_js_1 = require("../../hex/index.js");
const signerNostr_js_1 = require("./signerNostr.js");
const verify_js_1 = require("./verify.js");
class SignerNostrPrivateKey extends signerNostr_js_1.SignerNostr {
    constructor(client, privateKey) {
        super(client);
        this.privateKey = (0, index_js_1.hexFrom)(privateKey);
    }
    async connect() { }
    async isConnected() {
        return true;
    }
    async getNostrPublicKey() {
        return (0, index_js_1.hexFrom)(secp256k1_1.schnorr.getPublicKey(this.privateKey.slice(2)));
    }
    async signNostrEvent(event) {
        const pubkey = (await this.getNostrPublicKey()).slice(2);
        const eventHash = (0, verify_js_1.nostrEventHash)({ ...event, pubkey });
        const signature = secp256k1_1.schnorr.sign(eventHash, this.privateKey.slice(2));
        return {
            ...event,
            id: (0, index_js_1.hexFrom)(eventHash).slice(2),
            pubkey,
            sig: (0, index_js_1.hexFrom)(signature).slice(2),
        };
    }
}
exports.SignerNostrPrivateKey = SignerNostrPrivateKey;
